src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -13,44 +14,107 @@ namespace Microsoft.CodeAnalysis
+
+        private readonly object _collectionLock = new object();
 
         internal AdditionalSourcesCollection()
         {
-            _sourcesAdded = PooledDictionary<string, SourceText>.GetInstance();
+            _hintNames = PooledHashSet<string>.GetInstance();
+            _sourcesAdded = ArrayBuilder<GeneratedSourceText>.GetInstance();
         }
 
-        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> sources)
+        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> existingSources)
             : this()
         {
-            foreach (var source in sources)
-            {
-                _sourcesAdded.Add(source.HintName, source.Text);
-            }
+            _sourcesAdded.AddRange(existingSources);
         }
 
         public void Add(string hintName, SourceText source)
         {
-            _sourcesAdded.Add(hintName, source);
+            if (string.IsNullOrWhiteSpace(hintName))
+            {
+                throw new ArgumentNullException(nameof(hintName));
+            }
+
+            // allow any identifier character or [.-_ ()[]{}]
+            for (int i = 0; i < hintName.Length; i++)
+            {
+                char c = hintName[i];
+                if (!UnicodeCharacterUtilities.IsIdentifierPartCharacter(c)
+                    && c != '.'
+                    && c != '-'
+                    && c != '_'
+                    && c != ' '
+                    && c != '('
+                    && c != ')'
+                    && c != '['
+                    && c != ']'
+                    && c != '{'
+                    && c != '}')
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.HintNameInvalidChar, c, i), nameof(hintName));
+                }
+            }
+
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
+                }
+
+                _hintNames.Add(hintName);
+                _sourcesAdded.Add(new GeneratedSourceText(hintName, source));
+            }
         }
 
         public void RemoveSource(string hintName)
         {
-            _sourcesAdded.Remove(hintName);
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    _hintNames.Remove(hintName);
+                    for (int i = 0; i < _sourcesAdded.Count; i++)
+                    {
+                        // check the hashcodes, as thats the comparison we're using to put the names into _hintNames
+                        if (_sourcesAdded[i].HintName.GetHashCode() == hintName.GetHashCode())
```

Comment:
I considered making the GeneratedSourceTexts implement equals and hashcode in terms of the hintname, but they're only equal in the context of a single generator (which is equivalent to a single `AdditionalSourcesCollection`) 

I suppose we could use a custom comparer just in this class and check the hintName? That would allows us to get rid of the two collections and locks?

Reply:
This seems a bit complicated. Why not just use a `Dictionary<string, GeneratedSourceText>` instead of having parallel collections here? Could event use a `ConcurrentDictionary` and avoid all of the `lock` usage.

Reply:
The reason was that Dictionary isn't ordered. This ensures that we get syntax trees back in the order in which they were added. Presumably we could use an ordered dictionary and sort by the hintName, which would at least be consistent, but not the order they were added in. Do you think it matters, as long as its deterministic? 

---
In reply to: [435565199](https://github.com/dotnet/roslyn/pull/44768#discussion_r435565199) [](ancestors = 435565199)

Reply:
I ended up simplifying and just keeping an array builder. That way we'll keep the ordering as they're added, but don't need the locks of the second collection. It means `Remove()` and `Contains()` (and by extension `Add()`) are all O(N) operations now, as we have to check the whole collection for collisions. If it turns out to be a perf bottleneck we can fix it later on.

This also makes us not-thread safe as a generator could theoretically create a race condition that gets two additions into the collection with the same `hintName`. This could cause the compiler to crash further on (when adding to the PDBs for instance), but for now I think we're willing to live with that. I mean they could just `Environment.FailFast` if that was their aim.

---
In reply to: [435637087](https://github.com/dotnet/roslyn/pull/44768#discussion_r435637087) [](ancestors = 435637087,435565199)